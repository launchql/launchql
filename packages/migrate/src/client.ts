import { Pool, PoolConfig } from 'pg';
import { readFileSync } from 'fs';
import { join, dirname } from 'path';
import { Logger } from '@launchql/logger';
import { getPgPool } from 'pg-cache';
import { PgConfig } from 'pg-env';
import {
  MigrateConfig,
  DeployOptions,
  RevertOptions,
  VerifyOptions,
  DeployResult,
  RevertResult,
  VerifyResult,
  StatusResult
} from './types';
import { parsePlanFileSimple, Change, readScript, scriptExists } from '@launchql/core';
import { hashFile } from './utils/hash';
import { cleanSql } from './clean';
import { withTransaction, executeQuery, TransactionContext } from './utils/transaction';

// Helper function to get changes in order
function getChangesInOrder(planPath: string, reverse: boolean = false): Change[] {
  const plan = parsePlanFileSimple(planPath);
  return reverse ? [...plan.changes].reverse() : plan.changes;
}

function ensurePlanFile(where: string, planPath: string) {
  if (!planPath || !planPath.endsWith('.plan')) {
    throw new Error(`${where}: Plan file path is required, was given ${planPath}`);
  }
}

const log = new Logger('migrate');

export class LaunchQLMigrate {
  private pool: Pool;
  private pgConfig: PgConfig;
  private initialized: boolean = false;

  constructor(config: MigrateConfig) {
    this.pgConfig = {
      host: config.host,
      port: config.port,
      user: config.user,
      password: config.password,
      database: config.database
    };
    
    this.pool = getPgPool(this.pgConfig);
  }

  /**
   * Initialize the migration schema
   */
  async initialize(): Promise<void> {
    if (this.initialized) return;

    try {
      log.info('Checking LaunchQL migration schema...');
      
      // Check if launchql_migrate schema exists
      const result = await this.pool.query(`
        SELECT schema_name 
        FROM information_schema.schemata 
        WHERE schema_name = 'launchql_migrate'
      `);
      
      if (result.rows.length === 0) {
        log.info('Schema not found, creating migration schema...');
        
        // Read and execute schema SQL to create schema and tables
        const schemaPath = join(__dirname, 'sql', 'schema.sql');
        const proceduresPath = join(__dirname, 'sql', 'procedures.sql');
        
        const schemaSql = readFileSync(schemaPath, 'utf-8');
        const proceduresSql = readFileSync(proceduresPath, 'utf-8');
        
        await this.pool.query(schemaSql);
        await this.pool.query(proceduresSql);
        
        log.success('Migration schema created successfully');
      } else {
        log.success('Migration schema found and ready');
      }

      this.initialized = true;
    } catch (error) {
      log.error('Failed to initialize migration schema:', error);
      throw error;
    }
  }

  /**
   * Deploy changes according to plan file
   */
  async deploy(options: DeployOptions): Promise<DeployResult> {
    await this.initialize();
    
    const { project, targetDatabase, planPath, toChange, useTransaction = true } = options;
    ensurePlanFile('deploy', planPath);
    const plan = parsePlanFileSimple(planPath);
    const changes = getChangesInOrder(planPath);
    
    const deployed: string[] = [];
    const skipped: string[] = [];
    let failed: string | undefined;
    
    // Use a separate pool for the target database
    const targetPool = getPgPool({
      ...this.pgConfig,
      database: targetDatabase
    });
    
    // Execute deployment with or without transaction
    await withTransaction(targetPool, { useTransaction }, async (context) => {
      for (const change of changes) {
        // Stop if we've reached the target change
        if (toChange && deployed.includes(toChange)) {
          break;
        }
        
        // Check if already deployed using target database connection
        try {
          const deployedResult = await executeQuery(
            context,
            'SELECT launchql_migrate.is_deployed($1, $2) as is_deployed',
            [project || plan.project, change.name]
          );
          
          if (deployedResult.rows[0]?.is_deployed) {
            log.info(`Skipping already deployed change: ${change.name}`);
            skipped.push(change.name);
            continue;
          }
        } catch (checkError: any) {
          // If the function doesn't exist, the schema hasn't been initialized
          if (checkError.code === '42883') { // undefined_function
            log.debug('Migration schema not found, will be initialized on first deploy');
          } else {
            throw checkError;
          }
        }
        
        // Read deploy script
        const deployScript = readScript(dirname(planPath), 'deploy', change.name);
        if (!deployScript) {
          log.error(`Deploy script not found for change: ${change.name}`);
          failed = change.name;
          break;
        }

        const cleanDeploySql = await cleanSql(deployScript, false, '$EOFCODE$');
                
        // Calculate script hash
        const scriptHash = hashFile(join(dirname(planPath), 'deploy', `${change.name}.sql`));
        
        try {
          // Call the deploy stored procedure
          await executeQuery(
            context,
            'CALL launchql_migrate.deploy($1, $2, $3, $4, $5)',
            [
              project || plan.project,
              change.name,
              scriptHash,
              change.dependencies.length > 0 ? change.dependencies : null,
              cleanDeploySql
            ]
          );
          
          deployed.push(change.name);
          log.success(`Successfully deployed: ${change.name}`);
        } catch (error) {
          log.error(`Failed to deploy ${change.name}:`, error);
          failed = change.name;
          throw error; // Re-throw to trigger rollback if in transaction
        }
        
        // Stop if this was the target change
        if (toChange && change.name === toChange) {
          break;
        }
      }
    });
    
    return { deployed, skipped, failed };
  }

  /**
   * Revert changes according to plan file
   */
  async revert(options: RevertOptions): Promise<RevertResult> {
    await this.initialize();
    
    const { project, targetDatabase, planPath, toChange, useTransaction = true } = options;
    ensurePlanFile('revert', planPath);
    const plan = parsePlanFileSimple(planPath);
    const changes = getChangesInOrder(planPath, true); // Reverse order for revert
    
    const reverted: string[] = [];
    const skipped: string[] = [];
    let failed: string | undefined;
    
    // Use a separate pool for the target database
    const targetPool = getPgPool({
      ...this.pgConfig,
      database: targetDatabase
    });
    
    // Execute revert with or without transaction
    await withTransaction(targetPool, { useTransaction }, async (context) => {
      for (const change of changes) {
        // Stop if we've reached the target change
        if (toChange && change.name === toChange) {
          break;
        }
        
        // Check if deployed
        const deployedResult = await executeQuery(
          context,
          'SELECT launchql_migrate.is_deployed($1, $2) as is_deployed',
          [project || plan.project, change.name]
        );
        
        if (!deployedResult.rows[0]?.is_deployed) {
          log.info(`Skipping not deployed change: ${change.name}`);
          skipped.push(change.name);
          continue;
        }
        
        // Read revert script
        const revertScript = readScript(dirname(planPath), 'revert', change.name);
        if (!revertScript) {
          log.error(`Revert script not found for change: ${change.name}`);
          failed = change.name;
          break;
        }

        const cleanRevertSql = await cleanSql(revertScript, false, '$EOFCODE$');
        
        try {
          // Call the revert stored procedure
          await executeQuery(
            context,
            'CALL launchql_migrate.revert($1, $2, $3)',
            [project || plan.project, change.name, cleanRevertSql]
          );
          
          reverted.push(change.name);
          log.success(`Successfully reverted: ${change.name}`);
        } catch (error) {
          log.error(`Failed to revert ${change.name}:`, error);
          failed = change.name;
          throw error; // Re-throw to trigger rollback if in transaction
        }
      }
    });
    
    return { reverted, skipped, failed };
  }

  /**
   * Verify deployed changes
   */
  async verify(options: VerifyOptions): Promise<VerifyResult> {
    await this.initialize();
    
    const { project, targetDatabase, planPath } = options;
    ensurePlanFile('verify', planPath);
    const plan = parsePlanFileSimple(planPath);
    const changes = getChangesInOrder(planPath);
    
    const verified: string[] = [];
    const failed: string[] = [];
    
    // Use a separate pool for the target database
    const targetPool = getPgPool({
      ...this.pgConfig,
      database: targetDatabase
    });
    
    try {
      for (const change of changes) {
        // Check if deployed
        const isDeployed = await this.isDeployed(project, change.name);
        if (!isDeployed) {
          continue;
        }
        
        // Read verify script
        const verifyScript = readScript(dirname(planPath), 'verify', change.name);
        if (!verifyScript) {
          log.warn(`Verify script not found for change: ${change.name}`);
          continue;
        }
        
        const cleanVerifySql = await cleanSql(verifyScript, false, '$EOFCODE$');

        try {
          // Call the verify function
          const result = await targetPool.query(
            'SELECT launchql_migrate.verify($1, $2, $3) as verified',
            [project || plan.project, change.name, cleanVerifySql]
          );
          
          if (result.rows[0].verified) {
            verified.push(change.name);
            log.success(`Successfully verified: ${change.name}`);
          } else {
            failed.push(change.name);
            log.error(`Verification failed: ${change.name}`);
          }
        } catch (error) {
          log.error(`Failed to verify ${change.name}:`, error);
          failed.push(change.name);
        }
      }
    } finally {
    }
    
    return { verified, failed };
  }

  /**
   * Get deployment status
   */
  async status(project?: string): Promise<StatusResult[]> {
    await this.initialize();
    
    const result = await this.pool.query(
      'SELECT * FROM launchql_migrate.status($1)',
      [project]
    );
    
    return result.rows.map(row => ({
      project: row.project,
      totalDeployed: row.total_deployed,
      lastChange: row.last_change,
      lastDeployed: new Date(row.last_deployed)
    }));
  }

  /**
   * Check if a change is deployed
   */
  async isDeployed(project: string, changeName: string): Promise<boolean> {
    const result = await this.pool.query(
      'SELECT launchql_migrate.is_deployed($1, $2) as is_deployed',
      [project, changeName]
    );
    
    return result.rows[0].is_deployed;
  }

  /**
   * Check if Sqitch tables exist in the database
   */
  async hasSqitchTables(): Promise<boolean> {
    const result = await this.pool.query(`
      SELECT EXISTS (
        SELECT 1 FROM information_schema.tables 
        WHERE table_schema = 'sqitch' 
        AND table_name IN ('projects', 'changes', 'tags', 'events')
      )
    `);
    return result.rows[0].exists;
  }

  /**
   * Import from existing Sqitch deployment
   */
  async importFromSqitch(): Promise<void> {
    await this.initialize();
    
    try {
      log.info('Checking for existing Sqitch tables...');
      
      // Check if sqitch schema exists
      const schemaResult = await this.pool.query(
        "SELECT EXISTS (SELECT 1 FROM information_schema.schemata WHERE schema_name = 'sqitch')"
      );
      
      if (!schemaResult.rows[0].exists) {
        log.info('No Sqitch schema found, nothing to import');
        return;
      }
      
      // Import projects
      log.info('Importing Sqitch projects...');
      await this.pool.query(`
        INSERT INTO launchql_migrate.projects (project, created_at)
        SELECT DISTINCT project, now()
        FROM sqitch.projects
        ON CONFLICT (project) DO NOTHING
      `);
      
      // Import changes with dependencies
      log.info('Importing Sqitch changes...');
      await this.pool.query(`
        WITH change_data AS (
          SELECT 
            c.project,
            c.change,
            c.change_id,
            c.committed_at
          FROM sqitch.changes c
          WHERE c.change_id IN (
            SELECT change_id FROM sqitch.tags
          )
        )
        INSERT INTO launchql_migrate.changes (change_id, change_name, project, script_hash, deployed_at)
        SELECT 
          encode(sha256((cd.project || cd.change || cd.change_id)::bytea), 'hex'),
          cd.change,
          cd.project,
          cd.change_id,
          cd.committed_at
        FROM change_data cd
        ON CONFLICT (project, change_name) DO NOTHING
      `);
      
      // Import dependencies
      log.info('Importing Sqitch dependencies...');
      await this.pool.query(`
        INSERT INTO launchql_migrate.dependencies (change_id, requires)
        SELECT 
          c.change_id,
          d.dependency
        FROM launchql_migrate.changes c
        JOIN sqitch.dependencies d ON d.change_id = c.script_hash
        ON CONFLICT DO NOTHING
      `);
      
      log.success('Successfully imported Sqitch deployment history');
    } catch (error) {
      log.error('Failed to import from Sqitch:', error);
      throw error;
    }
  }

  /**
   * Get recent changes
   */
  async getRecentChanges(targetDatabase: string, limit: number = 10): Promise<any[]> {
    const targetPool = getPgPool({
      ...this.pgConfig,
      database: targetDatabase
    });

    try {
      const result = await targetPool.query(`
        SELECT 
          c.change_name,
          c.deployed_at,
          c.project
        FROM launchql_migrate.changes c
        ORDER BY c.deployed_at DESC NULLS LAST
        LIMIT $1
      `, [limit]);
      
      return result.rows;
    } catch (error) {
      log.error('Failed to get recent changes:', error);
      throw error;
    }
  }

  /**
   * Get pending changes (in plan but not deployed)
   */
  async getPendingChanges(planPath: string, targetDatabase: string): Promise<string[]> {
    const plan = parsePlanFileSimple(planPath);
    const allChanges = getChangesInOrder(planPath);
    
    const targetPool = getPgPool({
      ...this.pgConfig,
      database: targetDatabase
    });

    try {
      const deployedResult = await targetPool.query(`
        SELECT c.change_name
        FROM launchql_migrate.changes c
        WHERE c.project = $1 AND c.deployed_at IS NOT NULL
      `, [plan.project]);
      
      const deployedSet = new Set(deployedResult.rows.map((r: any) => r.change_name));
      return allChanges.filter(c => !deployedSet.has(c.name)).map(c => c.name);
    } catch (error: any) {
      // If schema doesn't exist, all changes are pending
      if (error.code === '42P01') { // undefined_table
        return allChanges.map(c => c.name);
      }
      throw error;
    }
  }

  /**
   * Get all deployed changes for a project
   */
  async getDeployedChanges(targetDatabase: string, project: string): Promise<any[]> {
    const targetPool = getPgPool({
      ...this.pgConfig,
      database: targetDatabase
    });

    try {
      const result = await targetPool.query(`
        SELECT 
          c.change_name,
          c.deployed_at,
          c.script_hash
        FROM launchql_migrate.changes c
        WHERE c.project = $1 AND c.deployed_at IS NOT NULL
        ORDER BY c.deployed_at ASC
      `, [project]);
      
      return result.rows;
    } catch (error: any) {
      // If schema doesn't exist, no changes are deployed
      if (error.code === '42P01') { // undefined_table
        return [];
      }
      throw error;
    }
  }

  /**
   * Get dependencies for a change
   */
  async getDependencies(project: string, changeName: string): Promise<string[]> {
    await this.initialize();
    
    try {
      const result = await this.pool.query(
        `SELECT d.requires 
         FROM launchql_migrate.dependencies d
         JOIN launchql_migrate.changes c ON c.change_id = d.change_id
         WHERE c.project = $1 AND c.change_name = $2`,
        [project, changeName]
      );
      
      return result.rows.map(row => row.requires);
    } catch (error) {
      log.error(`Failed to get dependencies for ${project}:${changeName}:`, error);
      return [];
    }
  }

  /**
   * Close the database connection pool
   */
  async close(): Promise<void> {
    // Pool is managed by PgPoolCacheManager, no need to close
  }
}