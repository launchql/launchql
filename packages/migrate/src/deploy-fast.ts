import { resolve } from 'path';
import chalk from 'chalk';

import { LaunchQLOptions } from '@launchql/types';
import { getRootPgPool } from '@launchql/server-utils';
import { LaunchQLProject } from './class/launchql';
import { spawn } from 'child_process';
import { packageModule } from './package';

interface Extensions {
  resolved: string[];
  external: string[];
}

interface DeployFastOptions {
    opts: LaunchQLOptions;
    name: string;
    database: string;
    dir: string;
    usePlan: boolean;  
}

export const deployFast = async (
  options: DeployFastOptions
): Promise<Extensions> => {

  const {
    dir,
    name, 
    database,
    opts,
    usePlan
  } = options;

  const mod = new LaunchQLProject(dir);

  console.log(chalk.cyan(`\n🔍 Gathering modules from ${chalk.bold(dir)}...`));
  const modules = mod.getModuleMap();

  if (!modules[name]) {
    console.log(chalk.red(`❌ Module "${name}" not found in modules list.`));
    throw new Error(`Module "${name}" does not exist.`);
  }

  console.log(chalk.cyan(`📦 Resolving dependencies for ${chalk.bold(name)}...`));
  const extensions: Extensions = mod.getModuleExtensions();

  const pgPool = getRootPgPool({
    ...opts.pg,
    database
  });

  console.log(chalk.green(`\n🚀 Starting deployment to database ${chalk.bold(database)}...`));

  for (const extension of extensions.resolved) {
    try {
      if (extensions.external.includes(extension)) {
        const msg = `CREATE EXTENSION IF NOT EXISTS "${extension}" CASCADE;`;
        console.log(chalk.blue(`\n📥 Installing external extension: ${chalk.bold(extension)}`));
        console.log(chalk.gray(`> ${msg}`));
        await pgPool.query(msg);
      } else {
        const modulePath = resolve(mod.workspacePath, modules[extension].path);
        console.log(chalk.magenta(`\n📂 Deploying local module: ${chalk.bold(extension)}`));
        console.log(chalk.gray(`→ Path: ${modulePath}`));
        console.log(chalk.gray(`→ Command: sqitch deploy db:pg:${database}`));

        const project = new LaunchQLProject(modulePath);

        const pkg = packageModule(project.modulePath, {
            usePlan,
            extension: false
        });
        console.log(chalk.gray(`> ${pkg.sql}`));
        await pgPool.query(pkg.sql);
      }
    } catch (e) {
      console.log(chalk.red(`\n🛑 Error during deployment: ${e instanceof Error ? e.message : e}`));
      await pgPool.end();
      process.exit(1);
    }
  }

  console.log(chalk.green(`\n✅ Deployment complete for ${chalk.bold(name)}.\n`));
  await pgPool.end();
  return extensions;
};
