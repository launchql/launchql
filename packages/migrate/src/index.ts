export { LaunchQLMigrate } from './client';
export * from './types';
export { hashFile, hashString } from './utils/hash';
export { readScript, scriptExists } from '@launchql/config-files';
export { withTransaction, TransactionContext, TransactionOptions } from './utils/transaction';