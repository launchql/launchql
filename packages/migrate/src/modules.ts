import { readFileSync } from 'fs';
import { sync as glob } from 'glob';
import { basename, dirname, relative } from 'path';

export interface Module {
  path: string;
  requires: string[];
  version: string;
}

export type ModuleMap = Record<string, Module>;

/**
 * Parse a .control file and extract its metadata.
 */
const parseControlFile = (filePath: string, basePath: string): Module => {
  const contents = readFileSync(filePath, 'utf-8');
  const key = basename(filePath).split('.control')[0];
  const requires = contents
    .split('\n')
    .find((line) => /^requires/.test(line))
    ?.split('=')[1]
    .split(',')
    .map((req) => req.replace(/[\'\s]*/g, '').trim()) || [];

  const version = contents
    .split('\n')
    .find((line) => /^default_version/.test(line))
    ?.split('=')[1]
    .replace(/[\']*/g, '')
    .trim() || '';

  return {
    path: dirname(relative(basePath, filePath)),
    requires,
    version,
  };
};

/**
 * List all modules by parsing .control files in the provided directory.
 */
export const listModules = (workspaceDir: string): ModuleMap => {
  const moduleFiles = glob(`${workspaceDir}/**/*.control`).filter(
    (file: string) => !/node_modules/.test(file)
  );

  return moduleFiles.reduce<ModuleMap>((acc: ModuleMap, file: string) => {
    const module = parseControlFile(file, workspaceDir);
    acc[basename(file).split('.control')[0]] = module;
    return acc;
  }, {});
};

/**
 * Get the latest change from the sqitch.plan file for a specific module.
 */
export const latestChange = (
  sqlmodule: string,
  modules: ModuleMap,
  basePath: string
): string => {
  const module = modules[sqlmodule];
  if (!module) {
    throw new Error(`latestChange() ${sqlmodule} NOT FOUND!`);
  }

  const plan = readFileSync(`${basePath}/${module.path}/sqitch.plan`, 'utf-8')
    .split('\n')
    .map((line) => line.trim())
    .filter(Boolean);

  return plan[plan.length - 1].split(' ')[0];
};

/**
 * Get the latest change and version for a specific module.
 */
export const latestChangeAndVersion = (
  sqlmodule: string,
  modules: ModuleMap,
  basePath: string
): { change: string; version: string } => {
  const module = modules[sqlmodule];
  if (!module) {
    throw new Error(`latestChangeAndVersion() ${sqlmodule} NOT FOUND!`);
  }

  const plan = readFileSync(`${basePath}/${module.path}/sqitch.plan`, 'utf-8')
    .split('\n')
    .map((line) => line.trim())
    .filter(Boolean);

  const change = plan[plan.length - 1].split(' ')[0];
  const pkg = require(`${basePath}/${module.path}/package.json`);

  return { change, version: pkg.version };
};

/**
 * Get extensions and modules required by a specific module.
 */
export const getExtensionsAndModules = (
  sqlmodule: string,
  modules: ModuleMap
): { native: string[]; sqitch: string[] } => {
  const module = modules[sqlmodule];
  if (!module) {
    throw new Error(`getExtensionsAndModules() ${sqlmodule} NOT FOUND!`);
  }

  const native = module.requires.filter(
    (req) => !Object.keys(modules).includes(req)
  );
  const sqitch = module.requires.filter((req) =>
    Object.keys(modules).includes(req)
  );

  return { native, sqitch };
};

/**
 * Get extensions and modules with their latest changes and versions.
 */
export const getExtensionsAndModulesChanges = (
  sqlmodule: string,
  modules: ModuleMap,
  basePath: string
): {
  native: string[];
  sqitch: { name: string; latest: string; version: string }[];
} => {
  const { native, sqitch } = getExtensionsAndModules(sqlmodule, modules);

  const sqitchWithDetails = 
    sqitch.map( (mod) => {
      const { change, version } = latestChangeAndVersion(mod, modules, basePath);
      return { name: mod, latest: change, version };
    });

  return { native, sqitch: sqitchWithDetails };
};
