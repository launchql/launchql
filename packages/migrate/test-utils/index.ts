import { MigrateConfig } from '../src/types';
import { LaunchQLMigrate } from '../src/client';
import { mkdtempSync, rmSync, cpSync, writeFileSync, mkdirSync } from 'fs';
import { join, resolve } from 'path';
import { tmpdir } from 'os';
import { getPgPool, teardownPgPools } from 'pg-cache';
import { getPgEnvOptions, PgConfig } from 'pg-env';
import { Pool } from 'pg';

export const FIXTURES_PATH = resolve(__dirname, '../../../__fixtures__/migrate');

// Global teardown function to be called after all tests
export async function teardownAllPools(): Promise<void> {
  await teardownPgPools();
}

// Helper to close specific database pools
export async function closeDatabasePools(databases: string[]): Promise<void> {
  // For now, we'll rely on the global teardown
  // Individual database pool cleanup can cause issues with pg-cache
}

export interface TestDatabase {
  name: string;
  config: MigrateConfig;
  query(sql: string, params?: any[]): Promise<any>;
  exists(type: 'schema' | 'table', name: string): Promise<boolean>;
  getDeployedChanges(): Promise<any[]>;
  getDependencies(project: string, changeName: string): Promise<string[]>;
  close(): Promise<void>;
}

export interface Change {
  name: string;
  dependencies?: string[];
  timestamp?: string;
  planner?: string;
  email?: string;
  comment?: string;
}

export class MigrateTestFixture {
  private tempDirs: string[] = [];
  private databases: TestDatabase[] = [];
  private dbCounter = 0;
  private pools: Pool[] = [];

  async setupTestDatabase(): Promise<TestDatabase> {
    const dbName = `test_migrate_${Date.now()}_${Math.random().toString(36).substring(2, 8)}_${this.dbCounter++}`;
    
    // Get base config from environment using pg-env
    const baseConfig = getPgEnvOptions({
      database: 'postgres'
    });
    
    // Create database using admin pool
    const adminPool = getPgPool(baseConfig);
    await adminPool.query(`CREATE DATABASE "${dbName}"`);
    
    // Get config for the new test database
    const pgConfig = getPgEnvOptions({
      database: dbName
    });
    
    const config: MigrateConfig = {
      host: pgConfig.host,
      port: pgConfig.port,
      user: pgConfig.user,
      password: pgConfig.password,
      database: pgConfig.database
    };

    // Initialize migrate schema
    const migrate = new LaunchQLMigrate(config);
    await migrate.initialize();

    // Get pool for test database operations
    const pool = getPgPool(pgConfig);
    this.pools.push(pool);

    const db: TestDatabase = {
      name: dbName,
      config,
      
      async query(sql: string, params?: any[]) {
        return pool.query(sql, params);
      },

      async exists(type: 'schema' | 'table', name: string) {
        if (type === 'schema') {
          const result = await pool.query(
            `SELECT EXISTS (
              SELECT 1 FROM information_schema.schemata 
              WHERE schema_name = $1
            ) as exists`,
            [name]
          );
          return result.rows[0].exists;
        } else {
          const [schema, table] = name.includes('.') ? name.split('.') : ['public', name];
          const result = await pool.query(
            `SELECT EXISTS (
              SELECT 1 FROM information_schema.tables 
              WHERE table_schema = $1 AND table_name = $2
            ) as exists`,
            [schema, table]
          );
          return result.rows[0].exists;
        }
      },

      async getDeployedChanges() {
        const result = await pool.query(
          `SELECT project, change_name, deployed_at 
           FROM launchql_migrate.changes 
           ORDER BY deployed_at`
        );
        return result.rows;
      },

      async getDependencies(project: string, changeName: string) {
        const result = await pool.query(
          `SELECT d.requires 
           FROM launchql_migrate.dependencies d
           JOIN launchql_migrate.changes c ON c.change_id = d.change_id
           WHERE c.project = $1 AND c.change_name = $2`,
          [project, changeName]
        );
        return result.rows.map((row: any) => row.requires);
      },

      async close() {
        // Don't close the pool here as it's managed by pg-cache
        // Just mark this database as closed
      }
    };

    this.databases.push(db);
    return db;
  }

  setupFixture(fixtureName: string): string {
    const originalPath = join(FIXTURES_PATH, fixtureName);
    const tempDir = mkdtempSync(join(tmpdir(), 'migrate-test-'));
    const fixturePath = join(tempDir, fixtureName);
    
    cpSync(originalPath, fixturePath, { recursive: true });
    this.tempDirs.push(tempDir);
    
    return fixturePath;
  }

  createPlanFile(project: string, changes: Change[]): string {
    const tempDir = mkdtempSync(join(tmpdir(), 'migrate-test-'));
    this.tempDirs.push(tempDir);

    const lines = [
      '%syntax-version=1.0.0',
      `%project=${project}`,
      `%uri=https://github.com/test/${project}`,
      ''
    ];

    for (const change of changes) {
      let line = change.name;
      
      if (change.dependencies && change.dependencies.length > 0) {
        line += ` [${change.dependencies.join(' ')}]`;
      }
      
      line += ` ${change.timestamp || new Date().toISOString().replace(/\.\d{3}Z$/, 'Z')}`;
      line += ` ${change.planner || 'test'}`;
      line += ` <${change.email || 'test@example.com'}>`;
      
      if (change.comment) {
        line += ` # ${change.comment}`;
      }
      
      lines.push(line);
    }

    const planPath = join(tempDir, 'sqitch.plan');
    writeFileSync(planPath, lines.join('\n'));
    
    return tempDir;
  }

  createScript(dir: string, type: 'deploy' | 'revert' | 'verify', name: string, content: string): void {
    const scriptDir = join(dir, type);
    mkdirSync(scriptDir, { recursive: true });
    writeFileSync(join(scriptDir, `${name}.sql`), content);
  }

  async cleanup(): Promise<void> {
    // Close all test database connections first
    const dbNames = this.databases.map(db => db.name);
    
    // Close all pools for test databases
    await closeDatabasePools(dbNames);
    
    // Get admin pool for database cleanup
    // const adminConfig = getPgEnvOptions({
    //   database: 'postgres'
    // });
    // const adminPool = getPgPool(adminConfig);

    // Small delay to ensure connections are closed
    await new Promise(resolve => setTimeout(resolve, 50));

    for (const db of this.databases) {
      try {
        // Drop the database
        // await adminPool.query(`DROP DATABASE IF EXISTS "${db.name}"`);
      } catch (e) {
        // Ignore errors - database might have active connections
      }
    }

    // Remove temporary directories
    for (const dir of this.tempDirs) {
      try {
        rmSync(dir, { recursive: true, force: true });
      } catch (e) {
        // Ignore errors during cleanup
      }
    }
  }
}

// Helper class for building plan files programmatically
export class PlanBuilder {
  private project: string;
  private changes: Change[] = [];

  constructor(project: string) {
    this.project = project;
  }

  addChange(name: string, options: Partial<Change> = {}): this {
    this.changes.push({
      name,
      ...options
    });
    return this;
  }

  build(): { project: string; changes: Change[] } {
    return {
      project: this.project,
      changes: this.changes
    };
  }

  toString(): string {
    const lines = [
      '%syntax-version=1.0.0',
      `%project=${this.project}`,
      `%uri=https://github.com/test/${this.project}`,
      ''
    ];

    for (const change of this.changes) {
      let line = change.name;
      
      if (change.dependencies && change.dependencies.length > 0) {
        line += ` [${change.dependencies.join(' ')}]`;
      }
      
      line += ` ${change.timestamp || new Date().toISOString().replace(/\.\d{3}Z$/, 'Z')}`;
      line += ` ${change.planner || 'test'}`;
      line += ` <${change.email || 'test@example.com'}>`;
      
      if (change.comment) {
        line += ` # ${change.comment}`;
      }
      
      lines.push(line);
    }

    return lines.join('\n');
  }
}