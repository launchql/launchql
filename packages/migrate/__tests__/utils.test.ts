import { join } from 'path';

import { getExtensionsAndModules, getExtensionsAndModulesChanges, latestChange, listModules } from '../src/modules';
import { FIXTURES_PATH } from '../test-utils';

const PROJECT_PATH = join(FIXTURES_PATH, 'sqitch/launchql');

describe('sqitch modules', () => {
  it('should get modules', async () => {
    const modules = await listModules(PROJECT_PATH);
    expect(modules).toMatchSnapshot();
  });

  it('should get a moduleâ€™s last path', async () => {
    const modules = await listModules(PROJECT_PATH);
    const change = await latestChange('totp', modules, PROJECT_PATH);
    expect(change).toBe('procedures/generate_secret');
  });

  it('should create dependencies for cross-project requires', async () => {
    const modules = await listModules(PROJECT_PATH);
    const deps = await getExtensionsAndModules('utils', modules);
    expect(deps).toEqual({
      native: ['plpgsql', 'uuid-ossp'],
      sqitch: ['totp', 'pg-verify'],
    });
  });

  it('should create dependencies for cross-project requires with changes', async () => {
    const modules = await listModules(PROJECT_PATH);
    const deps = await getExtensionsAndModulesChanges('utils', modules, PROJECT_PATH);
    expect(deps).toMatchSnapshot();
  });
});
