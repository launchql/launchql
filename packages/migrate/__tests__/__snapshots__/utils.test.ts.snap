// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`sqitch modules should create dependencies for cross-project requires with changes 1`] = `
{
  "native": [
    "plpgsql",
    "uuid-ossp",
  ],
  "sqitch": [
    {
      "latest": "procedures/generate_secret",
      "name": "totp",
      "version": "0.0.1",
    },
    {
      "latest": "procedures/verify_view",
      "name": "pg-verify",
      "version": "0.0.1",
    },
  ],
}
`;

exports[`sqitch modules should get modules 1`] = `
{
  "pg-utilities": {
    "path": "packages/utilities",
    "requires": [
      "plpgsql",
    ],
    "version": "0.0.1",
  },
  "pg-verify": {
    "path": "packages/verify",
    "requires": [
      "plpgsql",
      "uuid-ossp",
      "pg-utilities",
    ],
    "version": "0.0.1",
  },
  "secrets": {
    "path": "packages/secrets",
    "requires": [
      "plpgsql",
      "uuid-ossp",
      "totp",
      "pg-verify",
    ],
    "version": "0.0.1",
  },
  "totp": {
    "path": "packages/totp",
    "requires": [
      "plpgsql",
      "uuid-ossp",
      "pgcrypto",
      "pg-verify",
    ],
    "version": "0.0.1",
  },
  "utils": {
    "path": "packages/utils",
    "requires": [
      "plpgsql",
      "uuid-ossp",
      "totp",
      "pg-verify",
    ],
    "version": "0.0.1",
  },
}
`;
